### PR TITLE
Improve error message for MismatchingTypes

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
NA

## Why are the changes needed?
The compilation error message is not clear to users. The error message only shows the node ID, so users don't know which task/workflow has problems.

Screenshot 2024-08-08 at 3 36 09 PM

## What changes were proposed in this pull request?
Add task/workflow name to the error message.

Screenshot 2024-08-08 at 1 22 42 PM

## How was this patch tested?
```bash
pyflyte run --remote flyte-example/improve_error/mismatching_type.py wf3
```

### Setup process
```python
import typing

from flytekit import task, workflow, ImageSpec

image = ImageSpec(registry="pingsutw", packages=["pandas", "mypy"])


@task(container_image=image)
def t1() -> int:
    return 3 + 2


@task(container_image=image)
def t2(a: str) -> int:
    print(a)
    return 3


@task()
def calculate_circle_circumference(radius: float) -> typing.Optional[float]:
    return 2 * 3.14 * radius  # Task to calculate the circumference of a circle


@task()
def calculate_circle_area(radius: float) -> typing.Optional[float]:
    return 3.14 * radius * radius  # Task to calculate the area of a circle


@task
def echo(a: float) -> typing.Optional[float]:
    return a


@workflow()
def subwf1() -> int:
    return t1()


@workflow()
def subwf2(a: str) -> int:
    return t2(a=a)


@workflow()
def wf():
    subwf2(a=t1())  # (promise) task -> workflow


@workflow()
def wf1():
    t2(a=t1())  # (promise) task -> task


@workflow()
def wf2():
    subwf2(a=subwf1())  # (promise) workflow -> workflow


@workflow()
def wf3():
    t2(a=subwf1())  # (promise) workflow -> task
```

### Screenshots

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
NA

## Docs link
NA

### CODE REVIEW

File: flyteadmin/pkg/workflowengine/impl/compiler.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ func (c *workflowCompiler) CompileWorkflow(
 
 	compiledWorkflowClosure, err := compiler.CompileWorkflow(primaryWf, subworkflows, tasks, launchPlans)
 	if err != nil {
-		return nil, errors.NewFlyteAdminErrorf(codes.InvalidArgument, "failed to compile workflow with err %v", err)
+		return nil, errors.NewFlyteAdminError(codes.InvalidArgument, err.Error())
```
Comment (pingsutw) [inline flyteadmin/pkg/workflowengine/impl/compiler.go:40]: To remove duplicate errors.

Screenshot 2024-08-08 at 3 19 53 PM

File: flytepropeller/pkg/compiler/validators/bindings.go [hunk 2]
```diff
@@ -152,7 +164,8 @@ func validateBinding(w c.WorkflowBuilder, nodeID c.NodeID, nodeParam string, bin
 					return param.GetType(), []c.NodeID{val.Promise.NodeId}, true
 				}
 
-				errs.Collect(errors.NewMismatchingTypesErr(nodeID, val.Promise.Var, sourceType.String(), expectedType.String()))
+				errs.Collect(errors.NewMismatchingTypesErr(node.GetId(), outputVar, sourceType.String(), inputVar, expectedType.String()))
+				return nil, nil, !errs.HasErrors()
```
Comment (eapolinario) [inline flytepropeller/pkg/compiler/validators/bindings.go:168]: is this right?

File: flytepropeller/pkg/compiler/errors/compiler_errors.go [hunk 3]
```diff
@@ -202,13 +202,17 @@ func NewDuplicateIDFoundErr(nodeID string) *CompileError {
...
-		nodeID,
-	)
+func NewMismatchingTypesErr(nodeID, fromVar, fromType, toVar, toType string) *CompileError {
+	var errMsg string
+	if len(toVar) == 0 {
+		errMsg = fmt.Sprintf("Variable [%v] (type [%v]) doesn't match expected type [%v].", fromVar, fromType,
+			toType)
+	} else {
+		errMsg = fmt.Sprintf("The output variable '%v' has type [%v], but it's assigned to"+
+			" the input variable '%v' which has type type [%v].", fromVar, fromType, toVar, toType)
```
Comment (eapolinario) [inline flytepropeller/pkg/compiler/errors/compiler_errors.go:212]: Should we have 2 different functions instead? One to handle the previous case and another one to handle the new case where `toVar` is set.

**Review by eapolinario — COMMENTED** [review]
This is easily one of the most useful PRs coming out of this effort to improve error messages. Thank you for working on it!

File: flytepropeller/pkg/compiler/validators/bindings.go (hunk 2 above)
Comment (pingsutw) [reply flytepropeller/pkg/compiler/validators/bindings.go:168]: yup, outputVar is a promise. https://github.com/flyteorg/flyte/blob/827d55b1b72cf6ca6051af13b3ff9dfc8f37a726/flytepropeller/pkg/compiler/validators/bindings.go#L126

File: flytepropeller/pkg/compiler/errors/compiler_errors.go (hunk 3 above)
Comment (pingsutw) [reply flytepropeller/pkg/compiler/errors/compiler_errors.go]: that's better. updated it, thanks~

File: flytepropeller/pkg/compiler/errors/compiler_errors.go [hunk 4]
```diff
@@ -205,8 +205,15 @@ func NewDuplicateIDFoundErr(nodeID string) *CompileError {
...
-			toType),
+		fmt.Sprintf("Variable [%v] (type [%v]) doesn't match expected type [%v].", fromVar, fromType, toType),
+		nodeID,
+	)
+}
+
+func NewMismatchingVariablesErr(nodeID, fromVar, fromType, toVar, toType string) *CompileError {
+	return newError(
+		MismatchingTypes,
+		fmt.Sprintf("The output variable '%v' has type [%v], but it's assigned to the input variable '%v' which has type type [%v].", fromVar, fromType, toVar, toType),
```
Comment (Future-Outlier) [inline flytepropeller/pkg/compiler/errors/compiler_errors.go:216]: > The output variable ... the input variable
The upstream variable ... the downstream variable
Do you think it's better?

**Comment by Future-Outlier** [comment]
I'm a little bit tired today, let me review this PR tomorrow morning when I'm energetic.

File: flytepropeller/pkg/compiler/errors/compiler_errors.go (hunk 4 above)
Comment (Future-Outlier) [reply flytepropeller/pkg/compiler/errors/compiler_errors.go:216]: Wait, I think the version now is better.
```bash
        Details: failed to compile workflow for [resource_type:WORKFLOW  project:"flytesnacks"  domain:"development"  name:"example.wf3"  version:"ArJxfyEQms9wkDQCL-2Tpw"] with err: Collected Errors: 1
        Error 0: Code: MismatchingTypes, Node Id: n1, Description: The output variable 'subwf1.o0' has type [simple:INTEGER], but it's assigned to the input variable 't2.a' which has type type [simple:STRING].
```

**Review by Future-Outlier — DISMISSED** [review]
LGTM, this is super helpful.

**Review by Future-Outlier — APPROVED** [review]

**Pull request merged**